src/util/ConcurrentCache.h

```diff
@@ -0,0 +1,296 @@
+   * the associated computeFunctions must yield the same results.
+   * @param computeFunction The actual computation. If the result has to be
+   * computed, computeFunction() is called.
+   * @return A shared_ptr to the computation result.
+   *
+   */
+  template <class ComputeFunction>
+  ResultAndCacheStatus computeOnce(const Key& key,
+                                   ComputeFunction computeFunction) {
+    return computeOnceImpl(false, key, std::move(computeFunction));
+  }
+
+  /// Similar to computeOnce, with the following addition: After the call
+  /// completes, the result will be pinned in the underlying cache.
+  template <class ComputeFunction>
+  ResultAndCacheStatus computeOncePinned(const Key& key,
+                                         ComputeFunction computeFunction) {
+    return computeOnceImpl(true, key, std::move(computeFunction));
+  }
+
+  /// Clear the cache (but not the pinned elements)
+  void clear() { _cacheAndInProgressMap.wlock()->_cache.clearUnpinnedOnly(); }
+
+  /// Clear the cache, including the pinned elements.
+  void clearAll() { _cacheAndInProgressMap.wlock()->_cache.clearAll(); }
+
+  /// The number of non-pinned elements in the cache
+  auto numCachedElements() const {
+    return _cacheAndInProgressMap.wlock()->_cache.numCachedElements();
+  }
+
+  /// The number of pinned elements in the underlying cache
+  auto numPinnedElements() const {
+    return _cacheAndInProgressMap.wlock()->_cache.numPinnedElements();
+  }
+
+  /// Total size of the non-pinned elements in the cache (the unit depends on
+  /// the cache's configuration)
+  auto cachedSize() const {
+    return _cacheAndInProgressMap.wlock()->_cache.cachedSize();
+  }
+
+  /// Total size of the non-pinned elements in the cache (the unit depends on
+  /// the cache's configuration)
+  auto pinnedSize() const {
+    return _cacheAndInProgressMap.wlock()->_cache.pinnedSize();
+  }
+
+  /// only for testing: get access to the implementation
+  auto& getStorage() { return _cacheAndInProgressMap; }
+
+  // is key in cache (not in progress), used for testing
+  bool cacheContains(const Key& k) const {
+    return _cacheAndInProgressMap.wlock()->_cache.contains(k);
+  }
+
+  // Get element from cache by its key.
+  // The behavior if the key is not present in the cache depends on the
+  // cache type's operator[]. The ad_utility caches will return a nullptr
+  // in this case (return type is shared_ptr<const Value> in this case)
+  auto resultAt(const Key& k) {
+    return _cacheAndInProgressMap.wlock()->_cache[k];
+  }
+
+ private:
+  using ResultInProgress = ConcurrentCacheDetail::ResultInProgress<Value>;
+
+  // We hold a cache, and a hashtable in which we store all the computations
+  // that are currently in progress (only finished results are added to the
+  // cache)
+  struct CacheAndInProgressMap {
+    Cache _cache;
+    // values that are currently being computed.
+    // the bool tells us whether this result will be pinned in the cache
```

Comment:
No newline and Sentence should always be properly capitalized.

// Map of cache entries (key-value pairs) currently being computed. The bool tells us whether this result will be pinned in the cache.